Build out scaffolding for Project Evaluate view

kevinearldenny opened the pull request:
## Overview

- Add new EvaluateMetric type
- Add ProjectEvaluateView
- Add ProjectEvaluateMetricDetail
- Build out actions for toggling evaluate mode and switching between metrics


### Checklist

- [x] Description of PR is in an appropriate section of `CHANGELOG.md` and grouped with similar changes, if possible

### Demo

![image](https://user-images.githubusercontent.com/66973361/110825262-dad89e00-8261-11eb-8b01-df5917368851.png)

![image](https://user-images.githubusercontent.com/66973361/110825304-e1671580-8261-11eb-8ff1-ed694e843958.png)


### Notes

Optional. Ancillary topics, caveats, alternative strategies that didn't work out, anything else.

## Testing Instructions

- Open a project
- Click the evaluate button in the header
- _Expect_: Evaluate view displayed, with metrics listed in sidebar
- Click a metric in the sidebar
- _Expect_: Metric detail view displayed

Closes #603

pcaisse reviewed (changes requested):
This is looking pretty good. Tested well.

I noticed this one warning in the console (looks like we need to set a `key` somewhere):
```
Warning: Each child in a list should have a unique "key" prop.

Check the render method of `ProjectEvaluateMetricDetail`. See https://fb.me/react-warning-keys for more information.
    in EmotionCssPropInternal (at ProjectEvaluateMetricDetail.tsx:89)
    in ProjectEvaluateMetricDetail (at ProjectEvaluateView.tsx:208)
    in div (created by Context.Consumer)
    in Styled(div) (created by ForwardRef)
    in ForwardRef (created by Context.Consumer)
    in EmotionCssPropInternal (at ProjectEvaluateView.tsx:143)
    in div (created by Context.Consumer)
    in Styled(div) (at ProjectEvaluateView.tsx:142)
    in ProjectEvaluateView (at ProjectScreen.tsx:195)
    in div (created by Context.Consumer)
    in Styled(div) (created by Context.Consumer)
```

I did noticed that the map is being reloaded when evaluate mode is toggled. I think we should try to avoid that if we could since it isn't strictly necessary. It looks like `ProjectEvaluateView` renders to `Map` in much the same way that `ProjectScreen` does. I think that instead of having wholly separate code paths for the rendering of `ProjectScreen`, it might make sense to add some conditional logic to hide/show the header if in evaluate mode and also conditionally show the project sidebar/evaluate sidebar depending on which mode we're in. That way we can avoid mounting/unmounting `Map` based on the mode, avoiding the refresh. It also should significantly reduce the number of props being fed into `ProjectEvaluateView` since most of them relate to populating the map correctly and allow for getting rid of the `setMap` callback in `ProjectEvaluateView` which is duplicated from `ProjectScreen`.

Let me know if that makes sense.

maurizi commented:
> I did notice that the map is being reloaded when evaluate mode is toggled. I think we should try to avoid that if we could since it isn't strictly necessary.

I would like to second this as well - Jeff wants to do an animation/transition on the map when you toggle evaluate mode, which requires not reloading the map.

kevinearldenny commented:
> This is looking pretty good. Tested well.
> 
> I noticed this one warning in the console (looks like we need to set a `key` somewhere):
> 
> ```
> Warning: Each child in a list should have a unique "key" prop.
> 
> Check the render method of `ProjectEvaluateMetricDetail`. See https://fb.me/react-warning-keys for more information.
>     in EmotionCssPropInternal (at ProjectEvaluateMetricDetail.tsx:89)
>     in ProjectEvaluateMetricDetail (at ProjectEvaluateView.tsx:208)
>     in div (created by Context.Consumer)
>     in Styled(div) (created by ForwardRef)
>     in ForwardRef (created by Context.Consumer)
>     in EmotionCssPropInternal (at ProjectEvaluateView.tsx:143)
>     in div (created by Context.Consumer)
>     in Styled(div) (at ProjectEvaluateView.tsx:142)
>     in ProjectEvaluateView (at ProjectScreen.tsx:195)
>     in div (created by Context.Consumer)
>     in Styled(div) (created by Context.Consumer)
> ```
> 
> I did notice that the map is being reloaded when evaluate mode is toggled. I think we should try to avoid that if we could since it isn't strictly necessary. It looks like `ProjectEvaluateView` renders to `Map` in much the same way that `ProjectScreen` does. I think that instead of having wholly separate code paths for the rendering of `ProjectScreen`, it might make sense to add some conditional logic to hide/show the header if in evaluate mode and also conditionally show the project sidebar/evaluate sidebar depending on which mode we're in. That way we can avoid mounting/unmounting `Map` based on the mode, avoiding the refresh. It also should significantly reduce the number of props being fed into `ProjectEvaluateView` since most of them relate to populating the map correctly and allow for getting rid of the `setMap` callback in `ProjectEvaluateView` which is duplicated from `ProjectScreen`.
> 
> Let me know if that makes sense.

Gotcha, this makes sense. I initially was thinking that we explicitly wanted a separate map view, since the container is of a different size and the data display is totally different, but in light of your note I refactored the `ProjectEvaluateView` component into a view that only lives in the sidebar. 

I also added `id` as a key prop to the list of features displayed in the `ProjectEvaluateMetricView` component.

pcaisse commented on src/shared/entities.d.ts:
```diff
@@ -15,6 +15,14 @@ export interface IUser {
+  readonly type?: string;
+  readonly count?: number;
+  readonly status?: boolean;
+  readonly adjective?: string;
```
I wonder if this should be called `description` or something since it's not really an adjective. It also does seem to be required so maybe could ditch the `?`.

pcaisse replied:
You could go ahead and constrain which types are possible here so that the compiler will call out any typos, like so:
```
export type EvaluateMetric = {
  readonly name: string;
  readonly status: boolean;
  readonly adjective: string;
  readonly type: "fraction" | "percent" | "count";
  readonly value: number; 
};
```
This can then be leveraged for exhaustiveness checking. Instead of as it is now where we return a nonsense value `???` if we make a mistake:
```
                      {metric.type === "fraction"
                        ? `${metric.count} / 18`
                        : metric.type === "percent"
                        ? `${metric.count}%`
                        : metric.type === "count"
                        ? `${metric.count}`
                        : "???"}
```
We can have a format function that will error at compile time if we try to feed it a bad value:
```ts
function formatMetricValue(metric: EvaluateMetric): string {
  switch (metric.type) {
    case "fraction":
      return `${metric.count} / 18`;
    case "percent":
      return `${metric.count}%`;
    case "count":
      return `${metric.count}`;
    default:
      const exhaustive: never = metric;
      return exhaustive;
  }
}
```

kevinearldenny replied:
@pcaisse This mostly does look like a nice improvement, but I'm getting a TS compile error:

```
src/client/components/ProjectEvaluateSidebar.tsx:132:15 - error TS2322: Type 'EvaluateMetric' is not assignable to type 'never'.

132         const exhaustive: never = metric;
```

Any thoughts on fixing this? I added a `// @ts-ignore` above it in the commit so it would compile, but then after it compiled it was still throwing an error that caused the whole screen to blank out when you attempted to navigate to Evaluate mode. I updated it to:

```
// @ts-ignore
const exhaustive: never = metric;
```

but there might be a better way to handle this?

pcaisse replied:
I think this would be a great thing to pair on if you'd like.

pcaisse reviewed (approved):
LGTM. Left a couple of TypeScript-ful comments.
